### PR TITLE
splices, snap() conversion, and tags

### PR DESCRIPTION
**Pull request by invisig0th**

### CODE REVIEW

File: synapse/lib/chop.py [hunk 1]
```diff
@@ -62,6 +64,16 @@ def hexstr(text):
 def onespace(text):
     return ' '.join(text.split())
 
+@s_cache.memoize(size=10000)
```
Comment (vEpiphyte) [inline synapse/lib/chop.py]: Is there a reason why you cannot use `@functools.lru_cache` here?

File: synapse/lib/chop.py [hunk 2]
```diff
@@ -62,6 +64,16 @@ def hexstr(text):
...
 def tag(text):
     text = text.lower().strip('#').strip()
     return '.'.join([onespace(t) for t in text.split('.')])
+
+@s_cache.memoize(size=10000)
```
Comment (vEpiphyte) [inline synapse/lib/chop.py]: Is there a reason why you cannot use `@functools.lru_cache` here?

File: synapse/lib/cache.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+'''
+A few speed optimized (lockless) cache helpers.  Use carefully.
+'''
+import collections
+
+import synapse.common as s_common
+
+def memoize(size=10000):
+
```
Comment (vEpiphyte) [inline synapse/lib/cache.py]: should have warning message here ;)

File: synapse/lib/node.py [hunk 4]
```diff
@@ -196,43 +203,54 @@ def getTag(self, name, defval=None):
...
         self.tags[name] = norm
-        self.xact.stor(sops)
-
-        if curv is s_common.novalu:
-            #TODO: splice
-            pass
+        self.snap.stor(sops)
 
+        # TODO: fire an onTagAdd handler...
+        self.snap.splice('tag:add', ndef=self.ndef, tag=name, valu=norm)
```
Comment (vEpiphyte) [inline synapse/lib/node.py]: tag add splices currently have the ival, if present, set to the `ival` key.  see https://github.com/vertexproject/synapse/blob/v0.0.55/synapse/cores/common.py#L1864

File: synapse/lib/snap.py [hunk 5]
```diff
@@ -411,22 +424,47 @@ def addNodes(self, nodedefs):
...
+    @contextlib.contextmanager
+    def bulkload(self):
 
-        self.layr._xactRunStors(self.xact, sops)
+        #mine = not self.bulk
+        #if mine:
+            #self.bulk = True
+            #self.bulksops = []
+
+        yield
```
Comment (vEpiphyte) [inline synapse/lib/snap.py]: yield None?

Comment (vEpiphyte) [reply synapse/lib/snap.py]: this feels incomplete

File: synapse/models/files.py [hunk 6]
```diff
@@ -1,107 +1,156 @@
...
+
+            return f'guid:{valu}', {}
+
+        if kind == 'sha256':
+
+            if len(valu) != 64:
+                raise s_exc.BadTypeValu(name=self.name, valu=valu)
+
+            s_common.uhex(valu)
+            return f'sha256:{valu}', {}
```
Comment (vEpiphyte) [inline synapse/models/files.py]: if we are assuming that the `sha256:` is a sha256, we should kick out a `sha:256` sub

File: synapse/models/files.py [hunk 7]
```diff
@@ -1,107 +1,156 @@
...
+            guid = s_common.guid()
+            norm = f'guid:{guid}'
+            return norm, {}
+
+        if valu.find(':') == -1:
+
+            # we're ok with un-adorned sha256s
+            if len(valu) == 64 and s_common.uhex(valu):
+                valu = valu.lower()
+                return f'sha256:{valu}', {}
```
Comment (vEpiphyte) [inline synapse/models/files.py]: if we are assuming that the `sha256:` is a sha256, we should kick out a `sha:256` sub

File: synapse/datamodel.py [hunk 8]
```diff
@@ -282,6 +282,9 @@ def __init__(self):
             ('created', ('time', {}), {'ro': 1,
                 'doc': 'The time the node was created in the cortex.',
             }),
+            ('seen', ('ival', {}), {
```
Comment (vEpiphyte) [inline synapse/datamodel.py]: This is untested

File: synapse/lib/node.py (hunk 4 above)
Comment (invisig0th) [reply synapse/lib/node.py]: For new tags, the ival *is* the valu so, this should be inclusive.

File: synapse/lib/snap.py (hunk 5 above)
Comment (invisig0th) [reply synapse/lib/snap.py]: Sure.  The explicit syntax is probably better.

File: synapse/models/files.py [hunk 9]
```diff
@@ -1,107 +1,156 @@
...
+
             raise s_exc.BadTypeValu(name=self.name, valu=valu)
 
-        s_common.uhex(valu)
+        kind, valu = valu.split(':')
```
[the remark this answers is not in the document]
Comment (invisig0th) [reply synapse/models/files.py]: good catch, thx.

File: synapse/models/files.py (hunk 6 above)
Comment (invisig0th) [reply synapse/models/files.py]: Oh, right.  Duh.  Thx.

File: synapse/tests/test_model_crypto.py [hunk 10]
```diff
@@ -13,7 +13,7 @@
                  'b0af2978802aabf35483bbfcc470d50d6afb4283c1d06d2bf27efe9d7c09f226895633a46c3d77173bf0db8634299462b5f' \
                  '29629ad3b0470c76ddfd331ed0207d4dbd5fd44a2f66ca5f802ac0130e4a4bb2c149b5baa7a373188823ee21fe2950a76c8' \
                  '18586919f7914453d'
-HEXSTR_PUBLIC_EXPONENT = '010001'
```
Comment (vEpiphyte) [inline synapse/tests/test_model_crypto.py]: amusingly, we should probably shovel in 0x10001 as an integer here ;)

**Pull request merged**